### PR TITLE
cclient/serializer/json:get_neighbors - serialize response

### PR DESCRIPTION
**Pull request by howjmay**
#872 
implement API get_neighbors response serialization

implement neighbor_info_utarray_to_json_array()

### CODE REVIEW

File: cclient/serialization/json/helpers.c [hunk 1]
```diff
@@ -83,6 +85,37 @@ retcode_t utarray_uint64_to_json_array(UT_array const* const ut,
   return RC_OK;
 }
 
+retcode_t neighbor_info_utarray_to_json_array(UT_array const* const ut,
+                                              cJSON* const json_root,
+                                              char const* const obj_name) {
+  if (utarray_len(ut) > 0) {
```
Comment (oopsmonk) [inline cclient/serialization/json/helpers.c]: check if `ut` is NULL.

File: cclient/serialization/json/tests/get_neighbors.c [hunk 2]
```diff
@@ -22,7 +22,49 @@ void test_serialize_get_neighbors(void) {
...
+      ",\"numberOfInvalidTransactions\":" STR(TEST_NEIGHBORS_NUMINVALIDTX1)
+      ",\"numberOfNewTransactions\":" STR(TEST_NEIGHBORS_NUMNEWTX1)
+      "},{\"address\":\"" TEST_NEIGHBORS_NEIGHBOR2
+      "\",\"numberOfAllTransactions\":" STR(TEST_NEIGHBORS_NUMALLTX2)
+      ",\"numberOfInvalidTransactions\":" STR(TEST_NEIGHBORS_NUMINVALIDTX2)
+      ",\"numberOfNewTransactions\":" STR(TEST_NEIGHBORS_NUMNEWTX2) "}]}";
+
+  get_neighbors_res_t* res = get_neighbors_res_new();
+  char_buffer_t* test_neighbors_neighbor1 = char_buffer_new();
+  char_buffer_allocate(test_neighbors_neighbor1,
```
Comment (oopsmonk) [inline cclient/serialization/json/tests/get_neighbors.c]: char_buffer_allocate is no need.

File: cclient/serialization/json/tests/get_neighbors.c [hunk 3]
```diff
@@ -22,7 +22,49 @@ void test_serialize_get_neighbors(void) {
...
+                       sizeof(TEST_NEIGHBORS_NEIGHBOR1));
+  char_buffer_set(test_neighbors_neighbor1, TEST_NEIGHBORS_NEIGHBOR1);
+  char_buffer_t* test_neighbors_neighbor2 = char_buffer_new();
+  char_buffer_allocate(test_neighbors_neighbor2,
```
Comment (oopsmonk) [inline cclient/serialization/json/tests/get_neighbors.c]: ditto

File: cclient/serialization/json/get_neighbors.c [hunk 4]
```diff
@@ -21,8 +21,42 @@ retcode_t json_get_neighbors_serialize_request(const serializer_t *const s,
...
+  }
+
+  ret = neighbor_info_utarray_to_json_array(obj, json_root, "neighbors");
+  if (ret) {
+    goto err;
+  }
+
+  json_text = cJSON_PrintUnformatted(json_root);
+  if (json_text) {
+    len = strlen(json_text);
```
Comment (oopsmonk) [inline cclient/serialization/json/get_neighbors.c]: you can use `char_buffer_set` instead of `char_buffer_allocate`

**Review by oopsmonk — CHANGES_REQUESTED** [review]

File: cclient/serialization/json/helpers.c [hunk 5]
```diff
@@ -83,6 +85,40 @@ retcode_t utarray_uint64_to_json_array(UT_array const* const ut,
   return RC_OK;
 }
 
+retcode_t neighbor_info_utarray_to_json_array(UT_array const* const ut,
```
Comment (oopsmonk) [inline cclient/serialization/json/helpers.c]: it should be a private method, move to `cclient/serialization/json/get_neighbors.c`

File: cclient/serialization/json/helpers.c [hunk 6]
```diff
@@ -83,6 +85,40 @@ retcode_t utarray_uint64_to_json_array(UT_array const* const ut,
...
+                                              cJSON* const json_root,
+                                              char const* const obj_name) {
+  if (!ut) {
+    return RC_CCLIENT_JSON_PARSE;
```
Comment (oopsmonk) [inline cclient/serialization/json/helpers.c]: RC_NULL_PARAM

**Review by oopsmonk — CHANGES_REQUESTED** [review]

File: cclient/serialization/json/get_neighbors.c [hunk 7]
```diff
@@ -9,20 +9,48 @@
...
+  ret = char_buffer_set(out, req_text);
+  return ret;
+}
+
+retcode_t json_get_neighbors_serialize_response(
+    serializer_t const *const s, get_neighbors_res_t const *const obj,
+    char_buffer_t *out) {
+  retcode_t ret = RC_OK;
+  const char *json_text = NULL;
+  size_t len;
```
Comment (oopsmonk) [inline cclient/serialization/json/get_neighbors.c]: unused variable

**Review by oopsmonk — CHANGES_REQUESTED** [review]

File: cclient/serialization/json/get_neighbors.c [hunk 8]
```diff
@@ -9,20 +9,81 @@
 #include "cclient/serialization/json/helpers.h"
 #include "cclient/serialization/json/logger.h"
 
-retcode_t json_get_neighbors_serialize_request(const serializer_t *const s,
+retcode_t neighbor_info_utarray_to_json_array(UT_array const *const ut,
```
Comment (thibault-martinez) [inline cclient/serialization/json/get_neighbors.c]: Make it static ?

File: cclient/serialization/json/get_neighbors.c [hunk 9]
```diff
@@ -9,20 +9,81 @@
...
+      cJSON_AddItemToArray(array_obj, json_nbr_info);
+    }
+  }
+  return RC_OK;
+}
+
+retcode_t json_get_neighbors_serialize_request(serializer_t const *const s,
                                                char_buffer_t *out) {
   retcode_t ret = RC_OK;
   const char *req_text = "{\"command\":\"getNeighbors\"}";
```
Comment (thibault-martinez) [inline cclient/serialization/json/get_neighbors.c]: const placement

File: cclient/serialization/json/get_neighbors.c [hunk 10]
```diff
@@ -9,20 +9,81 @@
...
+
+  ret = char_buffer_set(out, req_text);
+  return ret;
+}
+
+retcode_t json_get_neighbors_serialize_response(
+    serializer_t const *const s, get_neighbors_res_t const *const obj,
+    char_buffer_t *out) {
+  retcode_t ret = RC_OK;
+  const char *json_text = NULL;
```
Comment (thibault-martinez) [inline cclient/serialization/json/get_neighbors.c]: const placement

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

**Review by thibault-martinez — APPROVED** [review]

**Review by oopsmonk — APPROVED** [review]

**Comment by oopsmonk** [comment]
@HowJMay thanks :)

**Pull request merged**